compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs

```diff
@@ -24,6 +24,142 @@ use crate::hir_ty_lowering::{
 };
 
 impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
+    pub(crate) fn lower_where_predicates(
+        &self,
+        params: &[hir::GenericParam<'tcx>],
+        hir_predicates: &[hir::WherePredicate<'tcx>],
+        predicates: &mut FxIndexSet<(ty::Clause<'tcx>, Span)>,
+    ) {
+        let tcx = self.tcx();
+        // Add implicit predicates that should be treated as if the user has written them,
+        // including the implicit `T: Sized` for all generic parameters, and `ConstArgHasType`
+        // for const params.
+        for param in params {
+            match param.kind {
+                hir::GenericParamKind::Lifetime { .. } => (),
+                hir::GenericParamKind::Type { .. } => {
+                    // We shift the predicate by one because for a non-lifetime binder like,
+                    // `for<T> T: Trait`, we start out with a `T` with a debruijn index of `N`,
+                    // but we need to shift it to be `N + 1` when we construct a clause below.
+                    let param_ty = ty::shift_vars(self.tcx(), self.lower_ty_param(param.hir_id), 1);
+                    let mut bounds = Vec::new();
+                    // Params are implicitly sized unless a `?Sized` bound is found
+                    self.add_sized_bound(
+                        &mut bounds,
+                        param_ty,
+                        &[],
+                        Some((param.def_id, hir_predicates)),
+                        param.span,
+                    );
+                    trace!(?bounds);
+                    predicates.extend(bounds);
+                    trace!(?predicates);
+                }
+                hir::GenericParamKind::Const { .. } => {
+                    let param_def_id = param.def_id.to_def_id();
+                    let ct_ty = tcx.type_of(param_def_id).instantiate_identity();
+                    let ct = self.lower_const_param(param_def_id, param.hir_id);
+                    if let ty::ConstKind::Bound(..) = ct.kind() {
+                        // We don't allow const params in non-lifetime binders today.
+                        tcx.dcx().span_delayed_bug(
```

Comment:
ì guess that's why xd 🤔 why not?

Reply:
const eval is fucked up

Reply:
specifically, given a situation like: `for<const C: usize> [T; C]: Trait`, that `C` in the array type will turn into an anon const that has an escaping bound `^C`. If we were to support higher-ranked consts, then anon consts with bound consts in their scopes would need to insert additional early bound consts to capture them, kinda like how we do for late-bound regions and opaques.

Reply:
See: https://github.com/rust-lang/rust/pull/115486/files#diff-a4c015afff6dad42c5509797bcc09a33e07fa6c5cef506e6e02374aed368c14c

Reply:
Yep, if we are doing the mGCA style of lowering for bound argument then it ""should""[^1] just work. We probably still need to ban late-bound const params from any other anon consts that show up, but that can be a property of anon consts and not a property of binders.

[^1]: well, we probably need to make sure that other operations on `ty::Const` will handle bound consts correctly, but they probably do? idk.

Reply:
for gce its fucked though yeah :3